File: frontend/src/api/map.ts
import { apiRequest } from './client'

export interface StationTemplate {
  id: string
  osm_id: string | null
  name: string
  latitude: number
  longitude: number
  base_price: string
  metadata_json: Record<string, unknown>
}

export interface Refinery {
  id: string
  name: string
  latitude: number
  longitude: number
}

export interface RoadNode {
  id: string
  latitude: number
  longitude: number
}

export interface RoadEdge {
  id: string
  from_node_id: string
  to_node_id: string
  road_type: string
  is_closed: boolean
}

export interface TrafficLight {
  id: string
  road_node_id: string
  red_seconds: number
  yellow_seconds: number
  green_seconds: number
  offset_seconds: number
}

export interface MapData {
  stations: StationTemplate[]
  refineries: Refinery[]
  road_nodes: RoadNode[]
  road_edges: RoadEdge[]
  traffic_lights: TrafficLight[]
}

export function fetchMapData(): Promise<MapData> {
  return apiRequest<MapData>('/api/map')
}

File: backend/app/schemas/map.py
import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class StationTemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    osm_id: str | None
    name: str
    latitude: float
    longitude: float
    base_price: Decimal
    metadata_json: dict[str, object]


class RefineryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    latitude: float
    longitude: float


class RoadNodeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    latitude: float
    longitude: float


class RoadEdgeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    from_node_id: uuid.UUID
    to_node_id: uuid.UUID
    road_type: str
    is_closed: bool


class TrafficLightResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    road_node_id: uuid.UUID
    red_seconds: float
    yellow_seconds: float
    green_seconds: float
    offset_seconds: float


class MapDataResponse(BaseModel):
    stations: list[StationTemplateResponse]
    refineries: list[RefineryResponse]
    road_nodes: list[RoadNodeResponse]
    road_edges: list[RoadEdgeResponse]
    traffic_lights: list[TrafficLightResponse]
